x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java

```diff
@@ -310,7 +310,7 @@ public static void validateSignature(final SignedJWT jwt, final List<JWK> jwks)
                 return; // VERIFY SUCCEEDED
             }
         }
-        throw new Exception("Verify failed using " + jwksStrength.size() + " of " + jwks.size() + " provided JWKs.");
+        throw new InvalidSignatureException("Verify failed using " + jwksStrength.size() + " of " + jwks.size() + " provided JWKs.");
```

Comment:
I think you would need to check if jwksStrength is empty before the for-loop, and throw the original catch-all Exception. Otherwise the code will fall through an empty for-loop and report a misleading InvalidSignatureException.

There are 4 filters in the code before the for-loop.
1. JWT mandatory alg header vs mandatory realm allowed signature algorithms
2. JWT optional kid header vs optional JWK kid attributes
3. JWT mandatory alg header vs optional JWK alg attributes
4. JWT mandatory alg strength vs JWK strength (HMAC and EC only)

One example is the realm config may allow "HS512,ES512", but the submitted JWT was signed by HS256, HS384, ES256, or ES384.

Another example is realm config has a PKC JWKSet with one RSA key, and that one RSA JWK has attributes like kid=123 or alg=RS512, but the submitted JWT has a header like header kid=456 or alg=RS256.